zScanner/ViewControllers/Documents/DocumentsListViewController.swift

```diff
@@ -72,62 +74,125 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
-                        self.reloadDocumentTypes()
-                    }
+                    self.rightBarButtons = []
+                    self.handleError(error)
+                    
                 }
             })
             .disposed(by: disposeBag)
-    }
-    
-    @objc private func newDocument() {
-        self.coordinator.createNewDocument()
+        
+        departmentsViewModel.departmentState
+            .observeOn(MainScheduler.instance)
+            .subscribe(onNext: { [weak self] status in
+                guard let self = self else { return }
+                
+                switch status {
+                case .awaitingInteraction:
+                    self.departmentsViewModel.fetchDepartments()
+                    
+                case .loading:
+                    self.departmentsStackView.subviews.forEach({ $0.removeFromSuperview() })
+                    self.departmentsStackView.addArrangedSubview(self.departmentsLoadingView)
+                    
+                case .success:
+                    self.departmentsStackView.subviews.forEach({ $0.removeFromSuperview() })
+                    
+                    self.departmentsViewModel.departments.value
+                        .map({
+                            DepartmentButton(model: $0)
+                        })
+                        .forEach({ (button) in
+                            button.rx.tap
+                                .subscribe({ [weak self] _ in
+                                    self?.loadDocumentTypes(departmentCode: button.model.id)
+                                })
+                                .disposed(by: self.disposeBag)
+
+                            self.departmentsStackView.addArrangedSubview(button)
+                        })
+                    
+                case .error( _):
+                    self.departmentsStackView.subviews.forEach({ $0.removeFromSuperview() })
+                    
+                    self.departmentsStackView.addArrangedSubview(self.departmentsErrorView)
+                    
+                }
+            })
+            .disposed(by: disposeBag)
+        
+        departmentsErrorView.buttonTap
+            .subscribe(onNext: { [weak self] _ in
+                self?.departmentsViewModel.fetchDepartments()
+            })
+        .disposed(by: disposeBag)
     }
     
     @objc private func openMenu() {
         coordinator.openMenu()
     }
-    
-    @objc private func reloadDocumentTypes() {
-        viewModel.updateDocumentTypes()
+
+    private func loadDocumentTypes(departmentCode: String) {
+        documentsViewModel.fetchDocumentTypes(for: departmentCode)
     }
     
     private func setupView() {
         navigationItem.title = "document.screen.title".localized
         
-        view.addSubview(tableView)
-        tableView.snp.makeConstraints { make in
-            make.edges.equalToSuperview()
+        documentsTableView.dataSource = self
```

Comment:
why did you move the datasource setup from the lazy init here? I think it should go back. Or is there any specific reason?